File: scripts/generate_types.py
#!/usr/bin/env python3
"""Generate Pydantic v2 models from Alea's canonical JSON Schemas."""

from __future__ import annotations

import hashlib
import json
import keyword
import re
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
SCHEMA_DIR = ROOT / "shared" / "schemas"
OUTPUT = ROOT / "api" / "app" / "generated" / "schemas.py"
INIT_OUTPUT = OUTPUT.parent / "__init__.py"

Schema = dict[str, Any]
SchemaKey = tuple[str, str]


def pascal(value: str) -> str:
    parts = re.findall(r"[A-Za-z0-9]+", value)
    result = "".join(part[:1].upper() + part[1:] for part in parts) or "Anonymous"
    return f"Model{result}" if result[0].isdigit() else result


def field_name(value: str) -> str:
    result = re.sub(r"\W", "_", value)
    if not result or result[0].isdigit() or keyword.iskeyword(result):
        result = f"field_{result}"
    return result


def pointer_join(pointer: str, token: str) -> str:
    escaped = token.replace("~", "~0").replace("/", "~1")
    return f"{pointer}/{escaped}"


def pointer_get(document: Schema, pointer: str) -> Schema:
    if pointer in ("", "#"):
        return document
    current: Any = document
    for token in pointer.removeprefix("#/").split("/"):
        current = current[token.replace("~1", "/").replace("~0", "~")]
    return current


documents = {
    path.name: json.loads(path.read_text(encoding="utf-8"))
    for path in sorted(SCHEMA_DIR.glob("*.json"))
}
names: dict[SchemaKey, str] = {}
used_names: set[str] = set()


def unique_name(candidate: str, file_name: str) -> str:
    if candidate not in used_names:
        used_names.add(candidate)
        return candidate
    prefixed = pascal(Path(file_name).stem) + candidate
    index = 2
    while prefixed in used_names:
        prefixed = f"{pascal(Path(file_name).stem)}{candidate}{index}"
        index += 1
    used_names.add(prefixed)
    return prefixed


def is_named_schema(schema: Schema) -> bool:
    return bool(
        schema.get("type") == "object"
        or "properties" in schema
        or "oneOf" in schema
        or "allOf" in schema
        or schema.get("not") == {}
    )


def discover(file_name: str, schema: Schema, pointer: str, suggested: str) -> None:
    if is_named_schema(schema):
        candidate = pascal(str(schema.get("title") or suggested))
        names.setdefault((file_name, pointer), unique_name(candidate, file_name))
        suggested = names[(file_name, pointer)]

    for key, child in schema.get("$defs", {}).items():
        discover(file_name, child, pointer_join(pointer, "$defs") + f"/{key}", f"{suggested}{pascal(key)}")
    for key, child in schema.get("properties", {}).items():
        if isinstance(child, dict):
            discover(file_name, child, pointer_join(pointer_join(pointer, "properties"), key), f"{suggested}{pascal(key)}")
    items = schema.get("items")
    if isinstance(items, dict):
        discover(file_name, items, pointer_join(pointer, "items"), f"{suggested}Item")
    for keyword_name in ("oneOf", "anyOf", "allOf"):
        for index, child in enumerate(schema.get(keyword_name, [])):
            if isinstance(child, dict) and "$ref" not in child:
                discover(
                    file_name,
                    child,
                    pointer_join(pointer_join(pointer, keyword_name), str(index)),
                    f"{suggested}{pascal(keyword_name)}{index + 1}",
                )


for current_file, document in documents.items():
    discover(current_file, document, "#", pascal(Path(current_file).stem))


def resolve_ref(file_name: str, ref: str) -> tuple[str, str, Schema]:
    target_file_part, separator, fragment = ref.partition("#")
    target_file = Path(target_file_part).name if target_file_part else file_name
    pointer = f"#{fragment}" if separator else "#"
    return target_file, pointer, pointer_get(documents[target_file], pointer)


def constraint_args(schema: Schema) -> list[str]:
    mapping = {
        "minimum": "ge",
        "maximum": "le",
        "exclusiveMinimum": "gt",
        "exclusiveMaximum": "lt",
        "minLength": "min_length",
        "maxLength": "max_length",
        "minItems": "min_length",
        "maxItems": "max_length",
    }
    return [f"{target}={schema[source]!r}" for source, target in mapping.items() if source in schema]


def type_expr(file_name: str, schema: Schema, pointer: str) -> str:
    if not schema:
        return "Any"
    if "$ref" in schema:
        target_file, target_pointer, target = resolve_ref(file_name, schema["$ref"])
        return names.get((target_file, target_pointer), type_expr(target_file, target, target_pointer))
    if "const" in schema:
        return f"Literal[{schema['const']!r}]"
    if "enum" in schema:
        return "Literal[" + ", ".join(repr(value) for value in schema["enum"]) + "]"
    if "oneOf" in schema or "anyOf" in schema:
        variants = schema.get("oneOf", schema.get("anyOf", []))
        rendered = [
            type_expr(file_name, item, pointer_join(pointer_join(pointer, "oneOf"), str(index)))
            for index, item in enumerate(variants)
        ]
        return " | ".join(dict.fromkeys(rendered)) or "Any"
    if "allOf" in schema:
        referenced = [item for item in schema["allOf"] if "$ref" in item]
        if referenced:
            return type_expr(file_name, referenced[0], pointer)

    schema_type = schema.get("type")
    if isinstance(schema_type, list):
        non_null = [value for value in schema_type if value != "null"]
        base = type_expr(file_name, {**schema, "type": non_null[0]}, pointer) if non_null else "None"
        return f"{base} | None" if "null" in schema_type and base != "None" else base
    if schema_type == "null":
        return "None"
    if schema_type == "string":
        return {"uuid": "UUID", "date-time": "datetime"}.get(schema.get("format"), "str")
    if schema_type == "integer":
        return "int"
    if schema_type == "number":
        return "float"
    if schema_type == "boolean":
        return "bool"
    if schema_type == "array":
        item_schema = schema.get("items", {})
        item_pointer = pointer_join(pointer, "items")
        return f"list[{type_expr(file_name, item_schema, item_pointer)}]"
    if schema_type == "object" or "properties" in schema:
        return names.get((file_name, pointer), "dict[str, Any]")
    if schema.get("not") == {}:
        return "Never"
    return "Any"


def combined_object_parts(schema: Schema) -> tuple[list[Schema], list[Schema]]:
    bases: list[Schema] = []
    bodies: list[Schema] = []
    if schema.get("type") == "object" or "properties" in schema:
        bodies.append(schema)
    elif "allOf" in schema:
        for part in schema["allOf"]:
            (bases if "$ref" in part else bodies).append(part)
    else:
        bodies.append(schema)
    return bases, bodies


def render_object(file_name: str, pointer: str, schema: Schema, name: str) -> str:
    base_parts, body_parts = combined_object_parts(schema)
    bases = [type_expr(file_name, part, pointer) for part in base_parts]
    base = bases[0] if bases else "SchemaModel"
    properties: dict[str, Schema] = {}
    required: set[str] = set()
    for body in body_parts:
        properties.update(body.get("properties", {}))
        required.update(body.get("required", []))

    lines = [f"class {name}({base}):"]
    if not properties:
        if any(body.get("additionalProperties") is not False for body in body_parts):
            lines.append('    model_config = ConfigDict(extra="allow")')
        else:
            lines.append("    pass")
        return "\n".join(lines)

    for json_name, child in properties.items():
        child_pointer = pointer_join(pointer_join(pointer, "properties"), json_name)
        annotation = type_expr(file_name, child, child_pointer)
        args = constraint_args(child)
        python_name = field_name(json_name)
        if python_name != json_name:
            args.append(f"alias={json_name!r}")
        metadata: list[str] = []
        if args:
            metadata.append(f"Field({', '.join(args)})")
        if child.get("uniqueItems") is True:
            metadata.append("AfterValidator(_ensure_unique)")
        if metadata:
            annotation = f"Annotated[{annotation}, {', '.join(metadata)}]"
        if json_name not in required:
            if "None" not in annotation:
                annotation = f"{annotation} | None"
            lines.append(f"    {python_name}: {annotation} = None")
        else:
            lines.append(f"    {python_name}: {annotation}")

    title = schema.get("title")
    if title in {"BetProposal", "BetDebate"}:
        lines.extend(
            [
                "",
                '    @model_validator(mode="after")',
                "    def validate_decision_shape(self) -> Self:",
                '        if self.decision == "bet" and (self.plan is None or self.no_bet_reason is not None):',
                '            raise ValueError("bet requires plan and null no_bet_reason")',
                '        if self.decision == "no_bet" and (self.plan is not None or self.no_bet_reason is None):',
                '            raise ValueError("no_bet requires null plan and a reason")',
                "        return self",
            ]
        )
    if title == "MethodologyReview":
        lines.extend(
            [
                "",
                '    @model_validator(mode="after")',
                "    def validate_revision_shape(self) -> Self:",
                '        if self.decision == "revise_and_review" and not self.proposed_revision:',
                '            raise ValueError("revise_and_review requires proposed_revision")',
                '        if self.decision in {"support", "oppose"} and self.proposed_revision is not None:',
                '            raise ValueError("support and oppose require null proposed_revision")',
                "        return self",
            ]
        )
    return "\n".join(lines)


object_blocks: list[str] = []
root_blocks: list[str] = []
model_names: list[str] = []
for (current_file, pointer), name in names.items():
    schema = pointer_get(documents[current_file], pointer)
    if schema.get("type") == "object" or "properties" in schema or "allOf" in schema:
        object_blocks.append(render_object(current_file, pointer, schema, name))
        model_names.append(name)
    elif "oneOf" in schema or "anyOf" in schema:
        expression = type_expr(current_file, schema, pointer)
        root_blocks.append(f"class {name}(RootModel[{expression}]):\n    pass")
        model_names.append(name)

digest = hashlib.sha256(
    b"".join(path.read_bytes() for path in sorted(SCHEMA_DIR.glob("*.json")))
).hexdigest()
header = f'''# Generated by scripts/generate_types.py. DO NOT EDIT.\n# schema-sha256: {digest}\n\n'''
imports = '''from __future__ import annotations\n\nimport json\nfrom datetime import datetime\nfrom typing import Annotated, Any, Literal, Never, Self, TypeVar\nfrom uuid import UUID\n\nfrom pydantic import (\n    AfterValidator,\n    BaseModel,\n    ConfigDict,\n    Field,\n    RootModel,\n    model_validator,\n)\n\n\nUniqueValue = TypeVar("UniqueValue")\n\n\ndef _ensure_unique(items: list[UniqueValue]) -> list[UniqueValue]:\n    keys = [\n        json.dumps(\n            item.model_dump(mode="json") if isinstance(item, BaseModel) else item,\n            ensure_ascii=False,\n            sort_keys=True,\n            default=str,\n        )\n        for item in items\n    ]\n    if len(keys) != len(set(keys)):\n        raise ValueError("array items must be unique")\n    return items\n\n\nclass SchemaModel(BaseModel):\n    model_config = ConfigDict(extra="forbid")\n\n'''
rebuild = "\n\nfor _model in [\n" + "".join(f"    {name},\n" for name in model_names) + "]:\n"
rebuild += "    _model.model_rebuild(_types_namespace=globals())\n"

OUTPUT.parent.mkdir(parents=True, exist_ok=True)
OUTPUT.write_text(
    header + imports + "\n\n".join(object_blocks + root_blocks) + rebuild,
    encoding="utf-8",
)
INIT_OUTPUT.write_text(
    '"""Generated models backed by shared/schemas JSON Schema."""\n',
    encoding="utf-8",
)
print(f"generated {len(model_names)} Pydantic models -> {OUTPUT.relative_to(ROOT)}")

File: scripts/generate_types.ts
#!/usr/bin/env bun
/** Generate Zod schemas and inferred TypeScript types from canonical JSON Schema. */

import { createHash } from "node:crypto";
import { mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";

type JsonSchema = Record<string, unknown>;

const root = resolve(import.meta.dir, "..");
const schemaDir = join(root, "shared", "schemas");
const output = join(root, "web", "src", "generated", "schemas.ts");
const files = readdirSync(schemaDir).filter((name) => name.endsWith(".json")).sort();
const documents = new Map<string, JsonSchema>(
  files.map((name) => [name, JSON.parse(readFileSync(join(schemaDir, name), "utf8"))]),
);

function pascal(value: string): string {
  const result = (value.match(/[A-Za-z0-9]+/g) ?? ["Anonymous"])
    .map((part) => part.slice(0, 1).toUpperCase() + part.slice(1))
    .join("");
  return /^\d/.test(result) ? `Model${result}` : result;
}

function getPointer(document: JsonSchema, pointer: string): JsonSchema {
  if (!pointer || pointer === "#") return document;
  return pointer
    .replace(/^#\//, "")
    .split("/")
    .reduce<unknown>((current, token) => {
      const key = token.replaceAll("~1", "/").replaceAll("~0", "~");
      return (current as Record<string, unknown>)[key];
    }, document) as JsonSchema;
}

function resolveRef(currentFile: string, ref: string): [string, JsonSchema] {
  const [filePart, fragment] = ref.split("#", 2);
  const targetFile = filePart ? basename(filePart) : currentFile;
  const document = documents.get(targetFile);
  if (!document) throw new Error(`missing schema document for ${ref}`);
  return [targetFile, getPointer(document, fragment ? `#${fragment}` : "#")];
}

function literal(value: unknown): string {
  return `z.literal(${JSON.stringify(value)})`;
}

function applyConstraints(expression: string, schema: JsonSchema): string {
  let result = expression;
  if (typeof schema.minLength === "number") result += `.min(${schema.minLength})`;
  if (typeof schema.maxLength === "number") result += `.max(${schema.maxLength})`;
  if (typeof schema.minimum === "number") result += `.min(${schema.minimum})`;
  if (typeof schema.maximum === "number") result += `.max(${schema.maximum})`;
  if (typeof schema.exclusiveMinimum === "number") result += `.gt(${schema.exclusiveMinimum})`;
  if (typeof schema.exclusiveMaximum === "number") result += `.lt(${schema.exclusiveMaximum})`;
  if (typeof schema.minItems === "number") result += `.min(${schema.minItems})`;
  if (typeof schema.maxItems === "number") result += `.max(${schema.maxItems})`;
  if (schema.uniqueItems === true) result += ".superRefine(requireUniqueItems)";
  return result;
}

function render(fileName: string, schema: JsonSchema, seen: Set<string> = new Set()): string {
  if (Object.keys(schema).length === 0) return "z.unknown()";
  if (typeof schema.$ref === "string") {
    const key = `${fileName}:${schema.$ref}`;
    if (seen.has(key)) return "z.unknown()";
    const [targetFile, target] = resolveRef(fileName, schema.$ref);
    return render(targetFile, target, new Set([...seen, key]));
  }
  if ("const" in schema) return literal(schema.const);
  if (Array.isArray(schema.enum)) {
    const variants = schema.enum.map(literal);
    return variants.length === 1 ? variants[0] : `z.union([${variants.join(", ")}])`;
  }
  if (Array.isArray(schema.oneOf) || Array.isArray(schema.anyOf)) {
    const options = (schema.oneOf ?? schema.anyOf) as JsonSchema[];
    const variants = options.map((item) => render(fileName, item, seen));
    return variants.length === 1 ? variants[0] : `z.union([${variants.join(", ")}])`;
  }
  if (Array.isArray(schema.allOf) && schema.type !== "object" && !schema.properties) {
    const variants = (schema.allOf as JsonSchema[]).map((item) => render(fileName, item, seen));
    return variants.reduce((left, right) => `${left}.and(${right})`);
  }
  if (schema.not && Object.keys(schema.not as JsonSchema).length === 0) return "z.never()";

  const declaredTypes = Array.isArray(schema.type) ? (schema.type as string[]) : [schema.type];
  const nullable = declaredTypes.includes("null");
  const schemaType = declaredTypes.find((value) => value !== "null");
  if (!schemaType && nullable) return "z.null()";
  let expression: string;

  switch (schemaType) {
    case "string":
      expression = "z.string()";
      if (schema.format === "uuid") expression += ".uuid()";
      if (schema.format === "date-time") expression += ".datetime({ offset: true })";
      break;
    case "integer":
      expression = "z.number().int()";
      break;
    case "number":
      expression = "z.number()";
      break;
    case "boolean":
      expression = "z.boolean()";
      break;
    case "array":
      expression = `z.array(${render(fileName, (schema.items as JsonSchema) ?? {}, seen)})`;
      break;
    case "object": {
      const properties = (schema.properties as Record<string, JsonSchema> | undefined) ?? {};
      const required = new Set((schema.required as string[] | undefined) ?? []);
      if (Object.keys(properties).length === 0 && schema.additionalProperties !== false) {
        expression = "z.record(z.string(), z.unknown())";
        break;
      }
      const fields = Object.entries(properties).map(([name, child]) => {
        let childExpression = render(fileName, child, seen);
        if (!required.has(name)) childExpression += ".optional()";
        return `${JSON.stringify(name)}: ${childExpression}`;
      });
      expression = `z.object({${fields.join(", ")}})`;
      expression += schema.additionalProperties === false ? ".strict()" : ".passthrough()";
      break;
    }
    case "null":
      expression = "z.null()";
      break;
    default:
      expression = "z.unknown()";
  }

  expression = applyConstraints(expression, schema);
  if (schema.title === "BetProposal" || schema.title === "BetDebate") {
    expression += ".superRefine(requireBetDecisionShape)";
  }
  if (schema.title === "MethodologyReview") {
    expression += ".superRefine(requireMethodologyDecisionShape)";
  }
  return nullable && schemaType !== "null" ? `${expression}.nullable()` : expression;
}

const exports: Array<{ fileName: string; name: string; schema: JsonSchema }> = [];
const usedNames = new Set<string>();
for (const fileName of files) {
  const document = documents.get(fileName)!;
  const rootName = pascal(String(document.title ?? basename(fileName, ".json")));
  const uniqueRoot = usedNames.has(rootName) ? `${pascal(basename(fileName, ".json"))}${rootName}` : rootName;
  usedNames.add(uniqueRoot);
  exports.push({ fileName, name: uniqueRoot, schema: document });
  for (const [defName, defSchema] of Object.entries(
    (document.$defs as Record<string, JsonSchema> | undefined) ?? {},
  )) {
    const proposed = pascal(String(defSchema.title ?? `${pascal(basename(fileName, ".json"))}${pascal(defName)}`));
    const unique = usedNames.has(proposed) ? `${pascal(basename(fileName, ".json"))}${proposed}` : proposed;
    usedNames.add(unique);
    exports.push({ fileName, name: unique, schema: defSchema });
  }
}

const digest = createHash("sha256")
  .update(files.map((name) => readFileSync(join(schemaDir, name))).join(""))
  .digest("hex");
const blocks = exports.map(({ fileName, name, schema }) => {
  const schemaName = `${name}Schema`;
  return `export const ${schemaName} = ${render(fileName, schema)};\nexport type ${name} = z.infer<typeof ${schemaName}>;`;
});
const source = `// Generated by scripts/generate_types.ts. DO NOT EDIT.\n// schema-sha256: ${digest}\n\nimport { z } from "zod";\n\nfunction requireUniqueItems(items: unknown[], context: z.RefinementCtx): void {\n  const keys = items.map((item) => JSON.stringify(item));\n  if (new Set(keys).size !== keys.length) {\n    context.addIssue({ code: "custom", message: "array items must be unique" });\n  }\n}\n\nfunction requireBetDecisionShape(value: unknown, context: z.RefinementCtx): void {\n  if (typeof value !== "object" || value === null) return;\n  const data = value as Record<string, unknown>;\n  const invalidBet = data.decision === "bet" && (data.plan == null || data.no_bet_reason !== null);\n  const invalidNoBet = data.decision === "no_bet" && (data.plan !== null || typeof data.no_bet_reason !== "string");\n  if (invalidBet || invalidNoBet) {\n    context.addIssue({ code: "custom", message: "decision, plan, and no_bet_reason are inconsistent" });\n  }\n}\n\nfunction requireMethodologyDecisionShape(value: unknown, context: z.RefinementCtx): void {\n  if (typeof value !== "object" || value === null) return;\n  const data = value as Record<string, unknown>;\n  const invalidRevision = data.decision === "revise_and_review" && !(typeof data.proposed_revision === "string" && data.proposed_revision.length > 0);\n  const invalidFinal = (data.decision === "support" || data.decision === "oppose") && data.proposed_revision !== null;\n  if (invalidRevision || invalidFinal) {\n    context.addIssue({ code: "custom", message: "decision and proposed_revision are inconsistent" });\n  }\n}\n\n${blocks.join("\n\n")}\n`;

mkdirSync(dirname(output), { recursive: true });
writeFileSync(output, source, "utf8");
console.log(`generated ${exports.length} Zod schemas -> ${output.replace(`${root}/`, "")}`);
